pkg/controllers/rollout/controller.go

```diff
@@ -0,0 +1,278 @@
+	}
+	if err := r.Client.List(ctx, bindingList, crpLabelMatcher); err != nil {
+		klog.ErrorS(err, "Failed to list all the bindings associated with the clusterResourcePlacement",
+			"clusterResourcePlacement", klog.KObj(crp))
+		return nil, nil, controller.NewAPIServerError(false, err)
+	}
+	// classify the bindings into different categories
+	for idx := range bindingList.Items {
+		binding := bindingList.Items[idx]
+		switch binding.Spec.State {
+		case fleetv1beta1.BindingStateUnscheduled:
+			canBeReadyBindings = append(canBeReadyBindings, &binding)
+			if isBindingReady(&binding, readyTimeCutOff) {
+				readyBindings = append(readyBindings, &binding)
+			}
+			if binding.DeletionTimestamp == nil {
+				// it's not deleted yet, so it is a removal candidate
+				removeCandidates = append(removeCandidates, &binding)
+			} else {
+				// it can be deleted at any time, so it is an unavailable candidate
+				canBeUnavailableBindings = append(canBeUnavailableBindings, &binding)
+			}
+
+		case fleetv1beta1.BindingStateScheduled:
+			// the scheduler has picked a cluster for this binding
+			schedulerTargetedBinds = append(schedulerTargetedBinds, &binding)
+			// this binding has not been bound yet, so it is an update candidate
+			updateCandidates = append(updateCandidates, &binding)
+
+		case fleetv1beta1.BindingStateBound:
+			schedulerTargetedBinds = append(schedulerTargetedBinds, &binding)
+			canBeReadyBindings = append(canBeReadyBindings, &binding)
+			if isBindingReady(&binding, readyTimeCutOff) {
+				readyBindings = append(readyBindings, &binding)
+			}
+			// The binding needs update if it's not pointing to the latest resource snapshot
+			if binding.Spec.ResourceSnapshotName != latestResourceSnapshotName {
+				removeCandidates = append(removeCandidates, &binding)
+			}
+		}
+	}
+
+	// calculate the target number of bindings
+	targetNumber := 0
+	if crp.Spec.Policy.PlacementType == fleetv1beta1.PickAllPlacementType {
+		// we use the scheduler picked bindings as the target number since there is no target in the CRP
+		targetNumber = len(schedulerTargetedBinds)
+	} else {
+		// the CRP validation webhook should make sure this is not nil
+		targetNumber = int(*crp.Spec.Policy.NumberOfClusters)
+	}
+
+	// calculate the max number of bindings that can be unavailable according to user specified maxUnavailable
+	maxUnavailableNumber, _ := intstr.GetScaledValueFromIntOrPercent(crp.Spec.Strategy.RollingUpdate.MaxUnavailable, targetNumber, true)
+	minAvailableNumber := targetNumber - maxUnavailableNumber
+	// This is the lower bound of the number of bindings that can be available during the rolling update
+	// Since we can't predict the number of bindings that can be unavailable after they are applied, we don't take them into account
+	lowerBoundAvailable := len(readyBindings) - len(canBeUnavailableBindings)
+	for i := 0; i < lowerBoundAvailable-minAvailableNumber; i++ {
+		tobeRemovedBinding = append(tobeRemovedBinding, removeCandidates[i])
+	}
```

Comment:
if we still have some capacity (lowerBoundAvaiable - minAvailable - len(toBeRevmovedBinding)), we can update the resourceSnapshot on the bounded binding.